### ISSUE TITLE
Reassemble method called too late

### ISSUE THREAD

**Issue by rrousselGit** [flutter/flutter#26503]
Consider a widget that rebuilds very often (typically because of an animation). 
Then consider editing the sources of that widgets through hot-reload. 

It seems that when doing so, there's a chance for `build` method to be called _before_ `reassemble` was, but with the updated content. 

This is a big issue for the following open source library: https://github.com/rrousselGit/flutter_hooks
As it needs to override the `reassemble` behavior to handle potential changes within `build` method.

But if `build` method is called before `reassemble`, then the library won't be ready for it and the app will enter into an invalid state.

**Comment by rrousselGit** [comment]
That issue is critical for the sanity of https://github.com/rrousselGit/flutter_hooks

As such, I can work on it if needed. But I need some guidance, as I don't know where to start.

**Comment by Hixie** [comment]
We probably don't pause the VM while we're updating it, so there's a race between us finishing the update of the code, and us sending the message to do the reassemble. But fundamentally we can't have the VM be paused while we send it the message, because it's code running inside the VM that has to respond to the message...

We could send a "stop handling frame messages" message before we do the update, then do the update and schedule the reassemble, then re-enable handling of frame messages. That would need changes to the tool and the framework.

**Comment by rrousselGit** [comment]
I see, thanks for the explanations.

Is there anything I can do to fasten the fix, or any potential workaround?

This is the source of a fairly common bug when using the package listed above. That is because, with hooks we tend to _not_ use builders such as `AnimatedBuilder`, and do the listening directly on the widget needing it. 

That's fine but it means that with this issue, working on an animated widget while its animation is playing has about 50% chance to make the library throw.

**Comment by Hixie** [comment]
I've raised its priority, hopefully we can get it some attention.

If you wanted to push a PR to fix it that would certainly accelerate matters, but we will hopefully get to it soonish so that's certainly not required.

**Comment by dnfield** [comment]
@rrousselGit can you try the patch here https://github.com/flutter/flutter/compare/master...dnfield:reassemble_build?expand=1 ?

I believe it should fix things for you - but it's a little bit rough and needs some tests/refinement to really work.  Either way it would be good to know if this is doing what I think it does.

**Comment by rrousselGit** [comment]
No, I can still see `build` called before `reassemble` but with updated code. 

An easy way to test this is using `flutter_hooks` with the following widget:

```dart
class Test extends HookWidget {
  @override
  Widget build(BuildContext context) {
    final animationController =
        useAnimationController(duration: const Duration(hours: 2));
    // useMemoized(animationController.forward, [animationController]);
    final animation = useAnimation(animationController);

    return Text(animation.toString());
  }
}
```

and repeatedly commenting/uncommenting the `useMemoized` line. 

If there's a crash which looks like the following, then the issue is still happening:

```
I/flutter (29532): 'package:flutter_hooks/src/framework.dart': Failed assertion: line 357 pos 14:
I/flutter (29532): '_currentHook.current?.hook?.runtimeType == hook.runtimeType': is not true.
```

**Comment by dnfield** [comment]
Ahh I missed something.  Can you try this again?  I'm not able to reproduce your bug now.

If this is working for you locally, can you come up with a good way to test it?  If you can take it over and add appropriate tests it will likely get done more quickly than I have time for at the moment.  Otherwise I'll work on it as I'm able.

**Comment by rrousselGit** [comment]
It seems to work. Nice!
I don't know of a good way to test it though, sorry

**Comment by Hixie** [comment]
For the record, the suggestion I put in https://github.com/flutter/flutter/pull/26983 was:

> We could do that by having an endpoint that we set a breakpoint on, which we call before hot reload, then do the hot reload while the process is paused, then resume, and have that endpoint immediately schedule a microtask that runs reassemble.

**Comment by dnfield** [comment]
One thing I'm not clear on is whether setting a breakpoint will pause the isolate, which AFAICT we can't do because that would cause the source reloading to stop.

Looking into this some more.

**Comment by Hixie** [comment]
@jonahwilliams suggests that instead of doing a breakpoint, we could teach the VM to call a specific entrypoint after it does a hot reload, and we could trigger the reassemble from there.

Long ago I had a suggestion to do something similar with passing in the list of objects that had been updated, so that we would know what to reassemble. Might be worth looking at doing that at the same time maybe.

**Comment by dnfield** [comment]
So basically mark reassemble as a `vm:entry-point`, and make the VM call it after a reload?

**Comment by jonahwilliams** [comment]
Something like that, we might have to define a new/hidden reassemble entrypoint

**Comment by dnfield** [comment]
I spent some time probing that idea and I'm not sure it will work - e.g. that it will actually stop a timer from running or prevent any other tasks/microtasks that had been scheduled before we got there from pre-empting our attempt to reassemble.

I'm starting to think the best we can do is just send a message that hot reload is about to start, and change the documentation on reassemble to clarify that other tasks/microtasks may pre-empt it, but that you can listen for the new message we'll send if it's important to your use case.

In the case for this bug, flutter_hooks could listen for that message and pause animations during the hot reload to ensure it has a good state to work with on reassemble.

**Comment by dnfield** [comment]
Unfortunately need to push this out again.  I have another idea I can try but probably won't get to it by the current milestone on here.

**Comment by dnfield** [comment]
I'm going to try to talk offline with @Hixie about this some more, but I'm wondering if this is really going to be doable.

If we pause or otherwise set a breakpoint on the Isolate, the reload doesn't make progress.  From what I can tell, the reload test that involves breakpoints is doing something else.

I did modify the Dart VM reload logic to invoke something that calls me back so that reassemble gets called earlier.  That seems to make the bug a little less likely to happen (we call `reassembleApplication` earlier than we otherwise would), but I'm still able to make it happen with some regularity. That may still be worth doing (it appears to significantly speed up hot reload in local tests), but I don't think we have any way to safely or consistently stop timers - while stopping only animation addresses the immediate concern of this bug, it would still leave the documentation inaccurate for cases involving non-animation related timers.

@rrousselGit - what would you think of just short circuiting your logic in this case?  I'm fairly sure that you'd eventually become consistent - the timer calling build would get stopped when reassemble finally gets called, and another call to build would get you back in business.

**Comment by rrousselGit** [comment]
> what would you think of just short-circuiting your logic in this case

How would that work? Locking builds? 

That would work for me. My only requirement is that build doesn't get called before I get notified of an implementation change.

____


After another shot at that issue, in my situation, it's not as important as I originally thought. 
I did get multiple reports of peoples falling for that bug, but it's not on a critical level.

Considering that it happens only in development, if it's too difficult to fix I don't mind it being delayed.

**Comment by dnfield** [comment]
I guess the idea is that if the types don't match during build, you just bail out and wait for the "next" build, which will come after reassemble.

**Comment by dnfield** [comment]
I've opened https://dart-review.googlesource.com/c/sdk/+/107120/1 to try to improve how early we call reassemble, but with a timer that's firing often enough we still fail this race.  As originally mentioned, we can't pause the isolate (including setting a breakpoint), and disabling timers seems very unsafe.

Even disabling animations (e.g. telling the framework to stop pumping frames for the duration of the reload) ends up looking pretty bad.  You get a momentary jank/non-responsiveness that makes the reload feel slower than it is.

**Comment by rrousselGit** [comment]
> if the types don't match during build, you just bail out and wait for the "next" build

I don't mind, but I don't see how that would be implemented.
Keeping a reference on `runtimeType`/`build` and compare them with the previous value before each build?

**Comment by dnfield** [comment]
If your types don't match what you expected from `reassemble`, fizzle out the build method with somethng safe (but don't throw/assert anything - maybe just return some kind of placeholder widget). Eventually, reassemble will fire and get you straigthened out, and build will get called again and you should be back on the happy path.

**Comment by Hixie** [comment]
We talked about how to do this and our theory is that we can fix it like this:
* Have a service extension that looks like this:
```dart
  void foo() {
    debugger(); // from dart:developer, triggers a breakpoint
    reassemble();
  }
```
* When we want to hot reload, call this service extension, wait for the isolate to pause.
* Cause the hot reload, wait for it to finish.
* Resume the isolate (rather than triggering a reassemble).
* Wait for the reassemble (the original service extension) to finish.

**Comment by dnfield** [comment]
I was pretty sure I had tried that, but really I had just tried something very close to it.

I think I have another idea about how to make this work now.  The thing blocking me before was that I wouldn't make progress because:

- You can't await the call that ends up setting a debugger - it won't return until the isolate is resumed, and the code that would resume it is going to come after the future completes.
- We have logic to look for paused isolates and _skip_ calling reassemble on them, the idea being that someone else paused them and we shouldn't be resuming them. We'd have to shuffle that around to look for paused isolates earlier instead.

**Comment by dnfield** [comment]
I'll need to come up with some way to test this, and to test the change I'm making to vmservice.dart, but https://github.com/flutter/flutter/pull/35514/files is yet another attempt that seems like it might work.

I got mixed up earlier on this one because I mistakenly thought setting the breakpoint was failing the hot reload - it wasn't, it just was that I was either awaiting something I couldn't, or that the isolate wasn't getting marked as resumed even when it was, or some combination of those factors.

**Comment by fzyzcjy** [comment]
Hi, is there any updates?
